File: javascript/subscription-manager/src/storage.js
import { Client, Subscription } from './models.js';

export class InMemoryBackend {
  constructor() {
    this.clients = new Map();            // email → Client
    this.subscriptions = new Map();      // clientId → [Subscription]
  }

  async getClientByEmail(email) {
    return this.clients.get(email) || null;
  }

  async addClient(client) {
    this.clients.set(client.email, client);
  }

  async getSubscriptions(clientId) {
    return this.subscriptions.get(clientId) || [];
  }

  async addSubscription(sub) {
    const arr = this.subscriptions.get(sub.clientId) || [];
    arr.push(sub);
    this.subscriptions.set(sub.clientId, arr);
  }
}

File: subscription_manager/storage.py
from abc import ABC, abstractmethod
from typing import List, Optional
from .models import Client, Subscription

class StorageBackend(ABC):
    @abstractmethod
    def get_client_by_email(self, email: str) -> Optional[Client]:
        ...

    @abstractmethod
    def add_client(self, client: Client) -> None:
        ...

    @abstractmethod
    def get_subscriptions(self, client_id: str) -> List[Subscription]:
        ...

    @abstractmethod
    def add_subscription(self, subscription: Subscription) -> None:
        ...
